src/mon/OSDMonitor.cc

```diff
@@ -11392,6 +11392,105 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
     wait_for_commit(op, new Monitor::C_Command(mon, op, 0, rs,
 						  get_last_committed() + 1));
     return true;
+  } else if (prefix == "osd crush reweight-by-scaling-factor" ||
+             prefix == "osd crush test-reweight-by-scaling-factor") {
+
+    bool really = false;
+    cmd_getval(cmdmap, "yes_i_really_mean_it", really);
+
+    bool skip_metadata_lookup = false;
+    cmd_getval(cmdmap, "skip_metadata_lookup", skip_metadata_lookup);
+
+    bool dry_run = prefix == "osd crush test-reweight-by-scaling-factor";
+
+    if (skip_metadata_lookup)
+      ss << "reweighting all OSDs using scaling factor "
+         << g_conf().get_val<double>("osd_crush_scaling_factor") << ":";
+    else
+      ss << "reweighting all OSDs using scaling factor "
+         << g_conf().get_val<double>("osd_crush_scaling_factor")
+         << " based on device size metadata:";
+
+    if (dry_run)
+      ss << "This is a dry run, no changes will be made. "
+         << "Use 'test-reweight-by-scaling-factor -o newcrush' "
+         << "then 'crushdiff compare -c newcrush' "
+         << "to check if this will cause data movement.";
+
+    CrushWrapper newcrush = _get_pending_crush();
+
+    for (int id = 0; id < newcrush.get_max_devices(); ++id) {
+        if (!newcrush.get_item_weight(id)) {
+          // weight is 0, skip
+          continue;
+        }
+        double new_weight = g_conf().get_val<double>("osd_crush_scaling_factor") *
+                                   newcrush.get_item_weightf(id);
+
+        if (!skip_metadata_lookup) {
+          map<string,string> meta;
+          err = load_metadata(id, meta, nullptr);
+          if (err < 0) {
+            ss << "failed to load metadata for osd." << id
+               << ": " << cpp_strerror(err);
+            goto reply_no_propose;
+          }
+
+          // osd statfs.total is bluestore_bdev_size + bluefs_db_size
+          uint64_t total_size = 0;
+          auto p1 = meta.find("bluestore_bdev_size");
```

Comment:
This will need updating for any new store - perhaps in the future we should report a generic usable size from the object store in the same way that OSD::update_crush_location calculates it so you don't have to recreate it here. Kind of surprised this isn't in the osdmap already.

Reply:
The update_crush_location() pathway should probably report the actual device size in bytes.  The mon side can then record that and apply the scaling factor afterwards.  See my comment above, we probably don't want each OSD using a local config to scale size.